### PR TITLE
feat: Add keeper functions for oracle

### PR DESCRIPTION
**Pull request by justyntemme**
## Description

Add functions on the keeper for oracle.
----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [x] updated the relevant documentation or specification
- [x] reviewed "Files changed" and left comments if necessary
- [x] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

**Review by toteki — APPROVED** [review]

File: x/oracle/keeper.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	gogotypes "github.com/gogo/protobuf/types"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	paramstypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
```
Comment (alexanderbez) [inline x/oracle/keeper.go:14]: nit: lets have these grouped

File: x/oracle/keeper.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+	StakingKeeper types.StakingKeeper
+
+	distrName string
+}
+
+// NewKeeper constructs a new keeper for oracle
+func NewKeeper(cdc codec.BinaryCodec, storeKey sdk.StoreKey,
+	paramspace paramstypes.Subspace, accountKeeper types.AccountKeeper,
+	bankKeeper types.BankKeeper, distrKeeper types.DistributionKeeper,
+	stakingKeeper types.StakingKeeper, distrName string) Keeper {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:37]: nit: long functions like this should be broken down as:

```
func(
  arg1,
  arg2,
  ...,
) Keeper {
  // ...
}
```

```suggestion
func NewKeeper(cdc codec.BinaryCodec, storeKey sdk.StoreKey,
	paramspace paramstypes.Subspace, accountKeeper types.AccountKeeper,
	bankKeeper types.BankKeeper, distrKeeper types.DistributionKeeper,
	stakingKeeper types.StakingKeeper, distrName string) Keeper {
```

File: x/oracle/keeper.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+//-----------------------------------
+// ExchangeRate logic
```
Comment (alexanderbez) [inline x/oracle/keeper.go:67]: nit: remove these redundant comments 👍

File: x/oracle/types/asset.go [hunk 4]
```diff
@@ -12,3 +17,26 @@ const (
 	BlocksPerYear   = BlocksPerDay * 365
 	MicroUnit       = int64(1e6)
 )
+
+type (
+	// GenesisState - all oracle state that must be provided at genesis
+
+	// ExchangeRatePrevote - struct to store a validator's prevote on the rate of Luna in the denom asset
```
Comment (alexanderbez) [inline x/oracle/types/asset.go:24]: nit: can we cleanup the godocs to be like `ExchangeRatePrevote defines a structure to ...`. This includes all the types here 👍

File: x/oracle/keeper.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+func (k Keeper) ClearTobinTaxes(ctx sdk.Context) {
+	store := ctx.KVStore(k.storeKey)
+	iter := sdk.KVStorePrefixIterator(store, types.TobinTaxKey)
+	defer iter.Close()
+	for ; iter.Valid(); iter.Next() {
+		store.Delete(iter.Key())
+	}
+}
+
+// ValidateFeeder return the given feeder is allowed to feed the message or not
```
Comment (alexanderbez) [inline x/oracle/keeper.go:349]:
```suggestion
// ValidateFeeder returns the given feeder is allowed to feed the message or not.
```

File: x/oracle/keeper.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+func (k Keeper) ValidateFeeder(ctx sdk.Context, feederAddr sdk.AccAddress, validatorAddr sdk.ValAddress) error {
+	if !feederAddr.Equals(validatorAddr) {
+		delegate := k.GetFeederDelegation(ctx, validatorAddr)
+		if !delegate.Equals(feederAddr) {
+			return sdkerrors.Wrap(types.ErrNoVotingPermission, feederAddr.String())
+		}
+	}
+
+	// Check that the given validator exists
```
Comment (alexanderbez) [inline x/oracle/keeper.go:358]:
```suggestion
	// check that the given validator exists
```

File: x/oracle/keeper.go [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+
+		var tobinTax sdk.DecProto
+		k.cdc.MustUnmarshal(iter.Value(), &tobinTax)
+		if handler(denom, tobinTax.Dec) {
+			break
+		}
+	}
+}
+
+// ClearTobinTaxes clears tobin taxes
```
Comment (alexanderbez) [inline x/oracle/keeper.go:339]:
```suggestion
// ClearTobinTaxes clears tobin taxes.
```

File: x/oracle/keeper.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+}
+
+// SetTobinTax updates tobin tax for the denom
+func (k Keeper) SetTobinTax(ctx sdk.Context, denom string, tobinTax sdk.Dec) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.cdc.MustMarshal(&sdk.DecProto{Dec: tobinTax})
+	store.Set(types.GetTobinTaxKey(denom), bz)
+}
+
+// IterateTobinTaxes iterates rate over tobin taxes in the store
```
Comment (alexanderbez) [inline x/oracle/keeper.go:323]:
```suggestion
// IterateTobinTaxes iterates rate over tobin taxes in the store.
```

File: x/oracle/keeper.go [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+	}
+}
+
+//-----------------------------------
+// AggregateExchangeRateVote logic
+
+// GetAggregateExchangeRateVote retrieves an oracle prevote from the store
+func (k Keeper) GetAggregateExchangeRateVote(ctx sdk.Context,
+	voter sdk.ValAddress) (aggregateVote types.AggregateExchangeRateVote,
+	err error) {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:258]: nit: please cleanup formatting here.

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Thanks for keeping the PR small. Was straightforward to review.

Please note, as we merge in this code piecemeal, it is imperative that we groom each bit of logic for both correctness and linting/style. Please keep this in mind 👍 

Nice work.

File: x/oracle/keeper.go (hunk 9 above)
Comment (justyntemme) [reply x/oracle/keeper.go:256]: made changes, will let you resolve if you are good with proposed formatting

File: x/oracle/keeper.go [hunk 10]
```diff
@@ -282,8 +280,9 @@ func (k Keeper) DeleteAggregateExchangeRateVote(ctx sdk.Context, voter sdk.ValAd
 }
 
 // IterateAggregateExchangeRateVotes iterates rate over prevotes in the store
-func (k Keeper) IterateAggregateExchangeRateVotes(ctx sdk.Context, handler func(voterAddr sdk.ValAddress,
-	aggregateVote types.AggregateExchangeRateVote) (stop bool)) {
+func (k Keeper) IterateAggregateExchangeRateVotes(ctx sdk.Context,
+	handler func(voterAddr sdk.ValAddress,
+		aggregateVote types.AggregateExchangeRateVote) (stop bool)) {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:285]: This is difficult to read. Perhaps consider making `func(voterAddr sdk.ValAddress,
		aggregateVote types.AggregateExchangeRateVote) (stop bool)` a type alias.

File: x/oracle/keeper.go [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+		dp := sdk.DecProto{}
+		k.cdc.MustUnmarshal(iter.Value(), &dp)
+		if handler(denom, dp.Dec) {
+			break
+		}
+	}
+}
+
+//-----------------------------------
+// Oracle delegation logic
```
Comment (alexanderbez) [inline x/oracle/keeper.go:113]: These need to be removed.

File: x/oracle/keeper.go [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+
+// GetFeederDelegation gets the account address that the validator operator delegated oracle vote rights to
+func (k Keeper) GetFeederDelegation(ctx sdk.Context, operator sdk.ValAddress) sdk.AccAddress {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.GetFeederDelegationKey(operator))
+	if bz == nil {
+		// By default the right is delegated to the validator itself
```
Comment (alexanderbez) [inline x/oracle/keeper.go:120]:
```suggestion
		// by default the right is delegated to the validator itself
```

File: x/oracle/keeper.go [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+	store.Set(types.GetExchangeRateKey(denom), bz)
+}
+
+// DeleteExchangeRate deletes the consensus exchange rate of USD denominated in the denom asset from the store.
+func (k Keeper) DeleteExchangeRate(ctx sdk.Context, denom string) {
+	store := ctx.KVStore(k.storeKey)
+	store.Delete(types.GetExchangeRateKey(denom))
+}
+
+// IterateExchangeRates iterates over USD rates in the store
```
Comment (alexanderbez) [inline x/oracle/keeper.go:97]:
```suggestion
// IterateExchangeRates iterates over USD rates in the store.
```

File: x/oracle/keeper.go [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+		if handler(denom, dp.Dec) {
+			break
+		}
+	}
+}
+
+//-----------------------------------
+// Oracle delegation logic
+
+// GetFeederDelegation gets the account address that the validator operator delegated oracle vote rights to
```
Comment (alexanderbez) [inline x/oracle/keeper.go:115]:
```suggestion
// GetFeederDelegation gets the account address that the validator operator delegated oracle vote rights to.
```

File: x/oracle/keeper.go [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+	bz := store.Get(types.GetFeederDelegationKey(operator))
+	if bz == nil {
+		// By default the right is delegated to the validator itself
+		return sdk.AccAddress(operator)
+	}
+
+	return sdk.AccAddress(bz)
+}
+
+// SetFeederDelegation sets the account address that the validator operator delegated oracle vote rights to
```
Comment (alexanderbez) [inline x/oracle/keeper.go:127]:
```suggestion
// SetFeederDelegation sets the account address that the validator operator delegated oracle vote rights to.
```

File: x/oracle/keeper.go (hunk 9 above)
Comment (alexanderbez) [reply x/oracle/keeper.go:256]: I would remove the named return values to shorten the signature.

File: x/oracle/keeper.go [hunk 16]
```diff
@@ -0,0 +1,357 @@
...
+
+// SetFeederDelegation sets the account address that the validator operator delegated oracle vote rights to.
+func (k Keeper) SetFeederDelegation(ctx sdk.Context, operator sdk.ValAddress, delegatedFeeder sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.GetFeederDelegationKey(operator), delegatedFeeder.Bytes())
+}
+
+// IterateFeederDelegations iterates over the feed delegates and performs a callback function.
+func (k Keeper) IterateFeederDelegations(ctx sdk.Context,
+	handler func(delegator sdk.ValAddress, delegate sdk.AccAddress) (stop bool)) {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:133]: nit: can we move this to a type alias please to shorten the function signature.

File: x/oracle/types/asset.go [hunk 17]
```diff
@@ -12,3 +17,24 @@ const (
 	BlocksPerYear   = BlocksPerDay * 365
 	MicroUnit       = int64(1e6)
 )
+
+type (
+	// ExchangeRatePrevote -  Defines a structure to store a validator's prevote on the rate of USD in the denom asset
```
Comment (alexanderbez) [inline x/oracle/types/asset.go:22]: You missed updating these type's godocs 👍

File: x/oracle/keeper.go [hunk 18]
```diff
@@ -0,0 +1,357 @@
...
+		if handler(operator, missCounter.Value) {
+			break
+		}
+	}
+}
+
+// GetAggregateExchangeRatePrevote retrieves an oracle prevote from the store
+func (k Keeper) GetAggregateExchangeRatePrevote(ctx sdk.Context,
+	voter sdk.ValAddress) (aggregatePrevote types.AggregateExchangeRatePrevote,
+	err error) {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:197]: formatting is off here.

**Review by alexanderbez — CHANGES_REQUESTED** [review]

File: x/oracle/keeper.go (hunk 18 above)
Comment (justyntemme) [reply x/oracle/keeper.go:197]: Going to remove the named return values, will comb through for all examples of that as well.

File: x/oracle/types/asset.go (hunk 17 above)
Comment (justyntemme) [reply x/oracle/types/asset.go:22]: I had updated with the defines a struct to store validator, i think i did not understand what you were asking, can you clarify for me?

File: x/oracle/keeper.go [hunk 19]
```diff
@@ -0,0 +1,361 @@
...
+func (k Keeper) SetFeederDelegation(ctx sdk.Context, operator sdk.ValAddress, delegatedFeeder sdk.AccAddress) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.GetFeederDelegationKey(operator), delegatedFeeder.Bytes())
+}
+
+type IterateFeederDelegationHandler = func(delegator sdk.ValAddress, delegate sdk.AccAddress) (stop bool)
+
+// IterateFeederDelegations iterates over the feed delegates and performs a callback function.
+func (k Keeper) IterateFeederDelegations(ctx sdk.Context,
+	handler IterateFeederDelegationHandler) {
```
Comment (alexanderbez) [inline x/oracle/keeper.go:135]:
```suggestion
func (k Keeper) IterateFeederDelegations(
    ctx sdk.Context,
    handler IterateFeederDelegationHandler,
) {
```

File: x/oracle/types/asset.go [hunk 20]
```diff
@@ -12,3 +17,24 @@ const (
 	BlocksPerYear   = BlocksPerDay * 365
 	MicroUnit       = int64(1e6)
 )
+
+type (
+	// ExchangeRatePrevote defines a structure to store a validator's prevote on the rate of USD in the denom asset
```
Comment (alexanderbez) [inline x/oracle/types/asset.go:22]:
```suggestion
	// ExchangeRatePrevote defines a structure to store a validator's prevote on
	// the rate of USD in the denom asset.
```

File: x/oracle/types/asset.go [hunk 21]
```diff
@@ -12,3 +17,24 @@ const (
...
+	ExchangeRatePrevote struct {
+		Hash        VoteHash       `json:"hash"`  // Vote hex hash to protect centralize data source problem
+		Denom       string         `json:"denom"` // Ticker name of target fiat currency
+		Voter       sdk.ValAddress `json:"voter"` // Voter val address
+		SubmitBlock int64          `json:"submit_block"`
+	}
+
+	// ExchangeRateVote defines a structure to store a validator's vote on the rate of USD in the denom asset
```
Comment (alexanderbez) [inline x/oracle/types/asset.go:30]:
```suggestion
	// ExchangeRateVote defines a structure to store a validator's vote on the
	// rate of USD in the denom asset.
```

**Review by alexanderbez — APPROVED** [review]

**Pull request merged**